### PR TITLE
[1.7.4?] Add LONG_WEAPON bonus

### PR DESCRIPTION
**Pull request by GeorgeK1ng**
### LONG_WEAPON

The affected unit can perform a melee attack from two hexes away if there is exactly one empty hex between attacker and target (`attacker -> empty hex -> victim`).

- If attacker is adjacent to the defender, this bonus does **not** extend range and attack is resolved as normal melee.
- Attacks performed at two-hex range do **not** allow retaliation.

Requested by @edeksumo. Please do various testing regarding to this touch.

### CODE REVIEW

File: lib/battle/CBattleInfoCallback.cpp [hunk 1]
```diff
@@ -835,6 +887,43 @@ bool CBattleInfoCallback::battleCanTargetEmptyHex(const battle::Unit * attacker)
 	return false;
 }
 
+bool CBattleInfoCallback::isLongWeaponAttack(const battle::Unit * attacker, const battle::Unit * defender) const
+{
+	RETURN_IF_NOT_BATTLE(false);
+
+	if(!attacker || !defender || attacker->doubleWide())
+		return false;
```
Comment (IvanSavenko) [inline lib/battle/CBattleInfoCallback.cpp]: - null attacker and null defender - better throw. This is incorrect API usage by called.
- why double-wide units can't use long attacks?

File: lib/battle/CBattleInfoCallback.cpp [hunk 2]
```diff
@@ -835,6 +887,43 @@ bool CBattleInfoCallback::battleCanTargetEmptyHex(const battle::Unit * attacker)
...
+			{
+				middleHex = defenderHex.cloneInDirection(static_cast<BattleHex::EDir>(direction), false);
+				attackerHex = middleHex.cloneInDirection(static_cast<BattleHex::EDir>(direction), false);
+			}
+			catch(const std::out_of_range &)
+			{
+				continue;
+			}
+
+			if(attacker->coversPos(attackerHex) && battleGetUnitByPos(middleHex, false) == nullptr)
```
Comment (IvanSavenko) [inline lib/battle/CBattleInfoCallback.cpp]: - Do corpses block long attacks?
- Do map obstacles block long attacks? What about walls? Or spells, like forcefield?

File: lib/battle/CBattleInfoCallback.cpp [hunk 3]
```diff
@@ -835,6 +887,43 @@ bool CBattleInfoCallback::battleCanTargetEmptyHex(const battle::Unit * attacker)
...
+				return true;
```
Comment (IvanSavenko) [inline lib/battle/CBattleInfoCallback.cpp]: You can put everything in try...catch to make it easier to read

```suggestion
			try
			{
				BattleHex middleHex = defenderHex.cloneInDirection(static_cast<BattleHex::EDir>(direction), false);
				BattleHex attackerHex = middleHex.cloneInDirection(static_cast<BattleHex::EDir>(direction), false);
				if(attacker->coversPos(attackerHex) && battleGetUnitByPos(middleHex, false) == nullptr)
					return true;
			}
			catch(const std::out_of_range &)
			{
				continue;
			}

```

**Comment by Alucard648** [comment]
I think, long reach attack should allow retaliation, if attacked creature also has Long Reach bonus, or Ranged Retaliation.

File: lib/battle/CBattleInfoCallback.cpp [hunk 4]
```diff
@@ -748,28 +775,53 @@ bool CBattleInfoCallback::battleCanAttackHex(const BattleHexArray & availableHex
...
 		}
-		const battle::Unit * defender = battleGetUnitByPos(position, false); //Do not allow to target corpses when standing on them (a WALK_AND_SPELLCAST action)
-		if (defender && defender->isDead() && defender->coversPos(fromHex))
+		catch(const std::out_of_range &)
+		{
 			return false;
+		}
+
+		if (battleGetUnitByPos(middleHex, false) == nullptr && canAttackFrom(longAttackFrom))
+			return true;
```
Comment (IvanSavenko) [inline lib/battle/CBattleInfoCallback.cpp]: Copy-pasted code?

File: client/battle/BattleAnimationClasses.cpp [hunk 5]
```diff
@@ -296,7 +296,30 @@ ECreatureAnimType MeleeAttackAnimation::selectGroup(bool multiAttack)
...
+				{
+					mutPos = direction;
+					break;
+				}
+			}
+			catch(const std::out_of_range &)
+			{
+				continue;
+			}
+		}
```
Comment (IvanSavenko) [inline client/battle/BattleAnimationClasses.cpp:326]: Copy-pasted 3rd time

File: lib/battle/CBattleInfoCallback.cpp (hunk 1 above)
Comment (GeorgeK1ng) [reply lib/battle/CBattleInfoCallback.cpp]: Done

File: lib/battle/CBattleInfoCallback.cpp (hunk 3 above)
Comment (GeorgeK1ng) [reply lib/battle/CBattleInfoCallback.cpp]: Done

File: lib/battle/CBattleInfoCallback.cpp (hunk 4 above)
Comment (GeorgeK1ng) [reply lib/battle/CBattleInfoCallback.cpp]: Done

File: client/battle/BattleAnimationClasses.cpp (hunk 5 above)
Comment (GeorgeK1ng) [reply client/battle/BattleAnimationClasses.cpp:326]: Done

File: lib/battle/CBattleInfoCallback.cpp (hunk 2 above)
Comment (GeorgeK1ng) [reply lib/battle/CBattleInfoCallback.cpp]: Good questions, let's wait for requestor answer. Whole PR is initial try to bring new bonus for Fallen of the Depth expansion. 

There is also request for another BREAK_WALLS bonus. When we are done with this one I would like to try to add it too. But rather in separate PR after this one.

Comment (Alucard648) [reply lib/battle/CBattleInfoCallback.cpp]: Corpses, quicksand, landmines, moat - No block for long reach.
Walls, obstacles, force fields - Block long reach.

Comment (GeorgeK1ng) [reply lib/battle/CBattleInfoCallback.cpp]: PR updated, walls / obstacles block attack. Corpses allow it. Also when creature stays in position to perform LONG_WEAPON attack, it should be done without moving. And when creature needs to move, will stop 1 hex before attacked unit.

EDIT: Confirmed by @edeksumo

**Comment by edeksumo** [comment]
I tested the changes, and here’s how it works:
A unit with this bonus will, by default, stop one hex away from the target unit to take advantage of the lack of retaliation. However, if that’s not possible, it attacks in melee.
Obstacles, walls, and force fields block attacks from one hex away.
Corpses and spells like mine land do not block it.
I also tested edge cases, such as attacking at the limit of movement range and attacking between two different units, and everything works as intended.

File: test/game/CGameStateTest.cpp [hunk 6]
```diff
@@ -463,6 +463,19 @@ TEST_F(CGameStateTest, battleInterference)
...
+	{
+		hero->removeBonusesRecursive(
+			Selector::sourceTypeSel(BonusSource::HERO_SPECIAL)
+			.Or(Selector::sourceTypeSel(BonusSource::SECONDARY_SKILL))
+		);
+	};
+
+	resetHeroBattleBonuses(attacker);
+	resetHeroBattleBonuses(defender);
+
```
Comment (IvanSavenko) [inline test/game/CGameStateTest.cpp]: That explains why this test would randomly fail.

But bug is someplace else. Heroes should be Catherine and Lord Haart: https://github.com/vcmi/vcmi/blob/develop/test/testdata/MiniTest/objects.json

Neither should have such specialties or skills

File: lib/battle/CBattleInfoCallback.cpp [hunk 7]
```diff
@@ -41,6 +41,32 @@ static BattleHex lineToWallHex(int line) //returns hex with wall in given line (
...
+	}
+}
+
+bool isLongWeaponMiddleHexClear(const CBattleInfoCallback & callback, const BattleHex & middleHex)
+{
+	if(!middleHex.isValid())
+		return false;
+
+	if(callback.battleGetUnitByPos(middleHex, true) != nullptr) // corpses are allowed for LONG_WEAPON
+		return false;
```
Comment (IvanSavenko) [inline lib/battle/CBattleInfoCallback.cpp]: Seems to be redundant - accessibility already accounts for hexes with units

File: lib/battle/CBattleInfoCallback.cpp [hunk 8]
```diff
@@ -690,7 +727,28 @@ BattleHex CBattleInfoCallback::fromWhichHexAttack(const battle::Unit * attacker,
 	}
 	if (direction == BattleHex::TOP || direction == BattleHex::BOTTOM)
 		return BattleHex::INVALID;
-	return target.cloneInDirection(direction, false);
+
+	BattleHex adjacentAttackFrom = target.cloneInDirection(direction, false);
+
+	if(attacker->hasBonusOfType(BonusType::LONG_WEAPON) && !attacker->doubleWide())
```
Comment (IvanSavenko) [inline lib/battle/CBattleInfoCallback.cpp]: Why double-wide units can't use long weapon?

File: lib/battle/CBattleInfoCallback.cpp [hunk 9]
```diff
@@ -1624,7 +1733,7 @@ AttackableTiles CBattleInfoCallback::getPotentiallyAttackableHexes(
 		attackDirection = BattleHex::mutualPosition(attackOriginHex, defender->occupiedHex(defenderPos));
 
 	if (attackDirection == BattleHex::NONE)
-		throw std::runtime_error("!!!");
+		return at;
```
Comment (IvanSavenko) [inline lib/battle/CBattleInfoCallback.cpp]: Why was this changed?

File: docs/modders/Bonus/Bonus_Types.md [hunk 10]
```diff
@@ -662,6 +662,13 @@ Deprecated. Please use [MULTIHEX_UNIT_ATTACK](#multihex_unit_attack) instead wit
 Similar to `TWO_HEX_ATTACK_BREATH`, but affecting two additional hexes in a triangular formation from the target hex.
 Deprecated. Please use [MULTIHEX_UNIT_ATTACK](#multihex_unit_attack) instead with custom icon and description.
 
+### LONG_WEAPON
+
+The affected unit can perform a melee attack from two hexes away if there is exactly one empty hex between attacker and target (`attacker -> empty hex -> victim`).
```
Comment (IvanSavenko) [inline docs/modders/Bonus/Bonus_Types.md]: From what I see unit will always pick long attack if possible - there is no way to actually pick which mode to use, right?

```suggestion
The affected unit will always perform a melee attack from two hexes away as long as there is exactly one empty hex between attacker and target (`attacker -> empty hex -> victim`).
```

**Review by IvanSavenko — COMMENTED** [review]
- Can AI use this feature?
- Perhaps add button on right panel to disable such attack in UI, similar to attack-and-return of Harpies?

File: lib/battle/CBattleInfoCallback.cpp (hunk 7 above)
Comment (GeorgeK1ng) [reply lib/battle/CBattleInfoCallback.cpp]: Removed

File: lib/battle/CBattleInfoCallback.cpp (hunk 8 above)
Comment (GeorgeK1ng) [reply lib/battle/CBattleInfoCallback.cpp]: Removed doubleWide checks, it's leftover.

File: lib/battle/CBattleInfoCallback.cpp (hunk 9 above)
Comment (GeorgeK1ng) [reply lib/battle/CBattleInfoCallback.cpp]: Leftover from testing, reverted

File: docs/modders/Bonus/Bonus_Types.md (hunk 10 above)
Comment (GeorgeK1ng) [reply docs/modders/Bonus/Bonus_Types.md]: Done

**Comment by GeorgeK1ng** [comment]
Added button for disabling Long weapon attack. Not sure, but now AI should use it too. actionNoLongWeapon.png is just placeholder created from actionAttack.png until @edeksumo create better one.

File: test/game/CGameStateTest.cpp (hunk 6 above)
Comment (GeorgeK1ng) [reply test/game/CGameStateTest.cpp]: Will drop it then later if there are no other things to tweak.

**Comment by Alucard648** [comment]
Speaking of long weapon: What about UI buttons, if creature has both Strike and Return + Long Weapon abilities?

**Comment by GeorgeK1ng** [comment]
I have tried gameplay and seems good, but of course require more testing. Both 1hex and 2hex creatures can use this bonus - interesting for unicorns or argali attacks for example. Initial bonus icon added to Extras PR https://github.com/vcmi-mods/vcmi-extras/pull/90 also created new button icon for this PR based on @edeksumo image. Still needs to be improved, but for this PR it should be enough as it's not used anywhere at the moment.

image

image

image

image

**Comment by Alucard648** [comment]
What about bonus that allows creature with Long Weapon bonus retaliate to attacking creature with Long Weapon as well?

**Comment by GeorgeK1ng** [comment]
Definitely not for this PR, it's already big. Feel free to create issue for future.

**Comment by Vltirr** [comment]
Sorry to barge in but, since you are developing this, would you consider also implementing a LONG_REACH attack? 
This would be same as LONG_WEAPON but with a ranged attack. So it would be a ranged attack with distance 1 that allows movement before attacking, something suited for skirmisher kind of units.

**Comment by GeorgeK1ng** [comment]
Same as previous comment.

File: test/game/CGameStateTest.cpp (hunk 6 above)
Comment (Vltirr) [reply test/game/CGameStateTest.cpp]: https://github.com/vcmi/vcmi/issues/7249

File: lib/battle/PossiblePlayerBattleAction.h [hunk 11]
```diff
@@ -25,6 +25,7 @@ class PossiblePlayerBattleAction // actions performed at l-click
 
 		MOVE_STACK,
 		ATTACK,
+		ATTACK_WITHOUT_LONG_WEAPON,
```
Comment (IvanSavenko) [inline lib/battle/PossiblePlayerBattleAction.h]: Probably should be the opposite - attack *with* long weapon and active by default? Or how it works with Harpies?

If it stays as "attack without long weapon" then icon & button description should be adapted - e.g. add crossed circle to icon

File: server/battles/BattleFlowProcessor.cpp [hunk 12]
```diff
@@ -604,14 +604,32 @@ bool BattleFlowProcessor::tryMakeAutomaticActionOfMeleeUnit(const CBattleInfoCal
...
+				BattleHex attackFromHex = battle.fromWhichHexAttack(actingStack, targetHex, directionEnum);
+				if(availableHexes.contains(attackFromHex))
+					attackableHexes.insert(attackFromHex);
+			}
+		}
+
+		bool isReachable = !attackableHexes.empty();
+		if(!isReachable)
+			continue;
+
```
Comment (IvanSavenko) [inline server/battles/BattleFlowProcessor.cpp:627]: Why we no longer can use `getAttackableHexes`?

Comment (GeorgeK1ng) [reply server/battles/BattleFlowProcessor.cpp:627]: We moved away from using getAttackableHexes here because it is only a geometric helper (adjacent tiles around the defender), not a full melee-legality check.
For forced CPU melee we need the same legality as real attack resolution, including movement availability, obstacles, and LONG_WEAPON rules.
battleCanAttackHex(...) already applies those checks, so we use it for consistency in target filtering/selection.

File: lib/battle/PossiblePlayerBattleAction.h (hunk 11 above)
Comment (GeorgeK1ng) [reply lib/battle/PossiblePlayerBattleAction.h]: Harpies uses their RETURN_AFTER_STRIKE ability by default, and can't be disabled by their special button. But seems there is place to improve it too. Because there are visible both buttons (Harpies one and use meele attack) and seems, like it has no functionality. Not sure how to handle them right now.

image

Regarding to current ATTACK_WITHOUT_LONG_WEAPON icon in button, it will be done later by feature requestor, probably in separate PR as this new LONG_WEAPON is not used anywhere in public, right now.

Comment (IvanSavenko) [reply lib/battle/PossiblePlayerBattleAction.h]: Still, clear LONG_WEAPON_ATTACK sounds like better option that current inverted one. I mean, Harpies action is called ATTACK_AND_RETURN, and not ATTACK_WITHOUT_RETURN or something. New option should work similar to that.

Comment (GeorgeK1ng) [reply lib/battle/PossiblePlayerBattleAction.h]: Renamed ATTACK_WITHOUT_LONG_WEAPON to LONG_WEAPON_ATTACK

Comment (IvanSavenko) [reply lib/battle/PossiblePlayerBattleAction.h]: Also needs logic inversion, to make it in line with ATTACK_AND_RETURN.

Check how `PossiblePlayerBattleAction::ATTACK_AND_RETURN` works - it has higher priority (5) than generic walk-and-attack (6). So if unit has both, game will pick attack-and-return. As result Harpies will by default use attack-and-return, and to use attack without returning, player can pick generic attack button

Comment (GeorgeK1ng) [reply lib/battle/PossiblePlayerBattleAction.h]: Performed inversion, should be now align with ATTACK_AND_RETURN. Also added new action icon create by @edeksumo which fit other action icons design.

**Review by IvanSavenko — APPROVED** [review]
